src/sentry/integrations/issue_alert_image_builder.py

```diff
@@ -32,33 +43,51 @@ def __init__(self, group: Group, provider: ExternalProviderEnum) -> None:
             "provider": self.provider,
             "issue_category": self.group.issue_category,
         }
+        self.lock = locks.get(key=f"lock_{self.cache_key}", duration=10, name="issue_alert_image")
 
     def get_image_url(self) -> str | None:
         try:
-            image_url = None
             metrics.incr("chartcuterie.issue_alert.attempt", tags=self.tags)
-            if self.group.issue_type == PerformanceP95EndpointRegressionGroupType:
-                image_url = self._get_endpoint_regression_image_url()
-            elif self.group.issue_type == ProfileFunctionRegressionType:
-                image_url = self._get_function_regression_image_url()
-
-            if image_url:
-                metrics.incr("chartcuterie.issue_alert.success", tags=self.tags)
-                return image_url
+            image_url = cache.get(self.cache_key)
+            if image_url is None:
+                self.lock.blocking_acquire(initial_delay=1, timeout=30)
+                # Checking again in case another thread generated the image while
+                # this thread was acquiring the lock
+                image_url = cache.get(self.cache_key)
+                if image_url is None:
+                    if self.group.issue_type == PerformanceP95EndpointRegressionGroupType:
+                        image_url = self._get_endpoint_regression_image_url()
+                    elif self.group.issue_type == ProfileFunctionRegressionType:
+                        image_url = self._get_function_regression_image_url()
+        except UnableToAcquireLock:
+            # There is a chance that another thread generated the image
+            image_url = cache.get(self.cache_key)
+            if not image_url:
+                logger.warning(
+                    "issue_alert_chartcuterie_image.lock.failed",
+                    extra={"group_id": self.group.id},
+                )
         except Exception as e:
             logger.exception(
                 "issue_alert_chartcuterie_image.failed",
                 extra={"exception": e, "group_id": self.group.id},
             )
             sentry_sdk.capture_exception()
+        self.lock.release()
```

Comment:
Is the lock always locked? Looks like it's behind `if image_url is None` no?

Reply:
Yes, if the image is already in cache we don't need lock. lock.release does nothing if the lock is not acquired, it fails silently. I'm adding it at the end to avoid writing it in all the ifs and elses.